dial.go

```diff
@@ -5,6 +5,9 @@ import (
 	"net"
 )
 
+// BypassRestrictionContextKey is the key used to bypass a restricted network.
+type BypassRestrictionContextKey struct{}
```

Comment:
How about hiding a bit of this complexity under helper functions, for example:
```go
func WithRestrictedNetworkBypass(ctx context.Context) context.Context {
    return context.WithValue(ctx, bypassRestrictedNetworkKey{}, struct{}{})
}

func HasRestrictedNetworkBypass(ctx context.Context) bool {
    return ctx.Value(bypassRestrictedNetworkKey{}) != nil
}

type bypassRestrictedNetworkKey struct{}
```
This way developers don't have to question what values they should put in the context, which is irrelevant anyway since we're only testing for existence.